core/tracker/inclusion.go

```diff
@@ -0,0 +1,300 @@
+type submission struct {
+	Duty     core.Duty
+	Pubkey   core.PubKey
+	Data     core.SignedData
+	AttRoot  eth2p0.Root
+	Delay    time.Duration
+	Included bool
+}
+
+// block is a simplified block with its attestations.
+type block struct {
+	Slot         int64
+	Attestations map[eth2p0.Root]*eth2p0.Attestation
+}
+
+// supported duty types.
+var supported = map[core.DutyType]bool{
+	core.DutyAttester:        true,
+	core.DutyAggregator:      true,
+	core.DutyProposer:        true,
+	core.DutyBuilderProposer: true,
+	// TODO(corver) Add support for sync committee and exit duties
+}
+
+// inclusion tracks the inclusion of submitted duties.
+// It has a simplified API to allow for easy testing.
+type inclusion struct {
+	mu          sync.Mutex
+	submissions []submission
+
+	missedFunc      func(context.Context, submission)
+	attIncludedFunc func(context.Context, submission, block)
+}
+
+// Submitted is called when a duty is submitted to the beacon node.
+// It adds the duty to the list of submitted duties.
+func (i *inclusion) Submitted(duty core.Duty, pubkey core.PubKey, data core.SignedData, delay time.Duration) error {
+	if !supported[duty.Type] {
+		return nil
+	}
+
+	var (
+		attRoot eth2p0.Root
+		err     error
+	)
+	if duty.Type == core.DutyAttester {
+		att, ok := data.(core.Attestation)
+		if !ok {
+			return errors.New("invalid attestation")
+		}
+		attRoot, err = att.HashTreeRoot()
+		if err != nil {
+			return errors.Wrap(err, "hash attestation")
+		}
+	} else if duty.Type == core.DutyAggregator {
+		agg, ok := data.(core.SignedAggregateAndProof)
+		if !ok {
+			return errors.New("invalid aggregate and proof")
+		}
+		attRoot, err = agg.Message.Aggregate.HashTreeRoot()
+		if err != nil {
+			return errors.Wrap(err, "hash aggregate")
+		}
+	}
+
+	i.mu.Lock()
+	defer i.mu.Unlock()
+	i.submissions = append(i.submissions, submission{
+		Duty:    duty,
+		Pubkey:  pubkey,
+		Data:    data,
+		AttRoot: attRoot,
+		Delay:   delay,
+	})
+
+	return nil
+}
+
+// Trim removes all duties that are older than the specified slot.
+// It also calls the missedFunc for any duties that have not been included.
+func (i *inclusion) Trim(ctx context.Context, slot int64) {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
+	var remaining []submission
+	for _, sub := range i.submissions {
+		if sub.Duty.Slot > slot {
+			remaining = append(remaining, sub)
+			continue
+		}
+		if !sub.Included {
+			i.missedFunc(ctx, sub)
+		}
+	}
+	i.submissions = remaining
+}
+
+// CheckBlock checks whether the block includes any of the submitted duties.
+func (i *inclusion) CheckBlock(ctx context.Context, block block) {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
+	for j, sub := range i.submissions {
+		if sub.Included {
+			continue
+		}
+
+		switch sub.Duty.Type {
+		case core.DutyAttester, core.DutyAggregator:
+			_, ok := block.Attestations[sub.AttRoot]
+			if !ok {
+				continue
+			}
+			i.submissions[j].Included = true
+			i.attIncludedFunc(ctx, sub, block)
```

Comment:
we need to also check the case where our single unaggregated attestation is included as part of aggregated attestation. Here you are checking if our unaggregated attestation is included or not OR aggregated attestation submitted by us is included or not.

AttRoot of an unaggregated attestation will have aggregation bits with only one bit set. While we need to check for the aggregation bits of an aggregated block attestation at the validator committee index like we were doing earlier.

Reply:
good catch, added "aggregated" logging field